tsdb/index/postings.go

```diff
@@ -156,24 +156,11 @@ func (p *MemPostings) Delete(deleted map[uint64]struct{}) {
 		}
 		p.mtx.RUnlock()
 
-		// For each posting we first analyse whether the postings list is affected by the deletes.
-		// If yes, we actually reallocate a new postings list.
 		for _, l := range vals {
 			// Only lock for processing one postings list so we don't block reads for too long.
 			p.mtx.Lock()
 
-			found := false
-			for _, id := range p.m[n][l] {
-				if _, ok := deleted[id]; ok {
-					found = true
-					break
-				}
-			}
-			if !found {
```

Comment:
Why did you remove the short circuit? The short circuit doesn't allocate anymore, but still avoids extra append calls and assignment.

Reply:
@csmarchbanks Actually the short circuit still need to iterate the p.m[n][l] one loop and if `found==true`, it need to allocate and iterate another loop to fullfill `repl`. But if we reuse the `p.m[n][l]`, only one loop is needed and no allocation at all under any conditions.

However as I commented above, reuse the slice will result a sneaky bug. So I'll rollback this code change.

Reply:
This code is inside of a Lock, so reusing the slice should not cause a problem.

I will rephrase my concern, the previous code did a loop with only gets in order to quickly see if anything changed. The new code only does one loop, but that loop is more expensive since it also does an append for every element that was not deleted. I am wondering if you have any benchmarks to show that doing the extra operations is ok.

Reply:
The lock only protect `MemPostings.m`, but not the memory underlying it.

For example, if `MemPosting.m["foo"]["bar"]` is {1, 2, 3, 4, 5} and through `MemPosting.Get("foo", "bar")` we will get a `ListPostings` whose `list` field share the same memory with `MemPosting.m["foo"]["bar"]`.

Then `MemPosting.Delete()` delete some ref, `2, 4` for example. If we reuse the slice, the underlying memory will be {1, 3, 5, 4, 5}. Finally, the iterating of `ListPostings` will not be correct.

Reply:
:+1: That make sense. Also, that is an insidious potential bug that does not fail any tests.

Reply:
So I'll delete this part of slice reuse. 

And I'd like to add a test for this insidious case if necessary :)

Reply:
> And I'd like to add a test for this insidious case if necessary :)

Please do, or else I will in a separate PR. :) It doesn't look like `MemPostings.Delete` has any package level testing right now.

Reply:
Another PR will be fine.